server/src/main/java/org/elasticsearch/index/mapper/TsidExtractingIdFieldMapper.java

```diff
@@ -72,6 +84,51 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
 
         @Override
         public Query termQuery(Object value, SearchExecutionContext context) {
+            if (value instanceof BytesRef == false) {
+                return termsQuery(Arrays.asList(value), context);
+            }
+            String id = ((BytesRef) value).utf8ToString();
+            if (id.startsWith("test:")) {
+                id = id.substring("test:".length());
+                IndexRouting.ExtractFromSource indexRouting = (IndexRouting.ExtractFromSource) context.getIndexSettings().getIndexRouting();
+                byte[] suffix = indexRouting.idToSuffix(id);
+                if (suffix == null || suffix.length != 16) {
+                    return new MatchNoDocsQuery();
+                }
+                long timestamp = ByteUtils.readLongLE(suffix, 8);
+                return new TermQuery(new Term(NAME, new BytesRef(id))) {
```

Comment:
Would it be an option to encode the timestamp as a prefix of the id so that the optimization would work out-of-the-box without needing a custom query?

Reply:
Yeah. I'm certainly thinking about it.

Reply:
I put the timestamp last to get the most shared prefixes. Maybe a bad choice, but it's what I was doing then. We can change it.

I encoded the timestamp in little endian because we had a little endian method. I still have an open follow up to evaluate flipping it.

Reply:
A (very late) update - we care a *lot* about the size of the `_id` now. And we think that an encoding scheme like @jpountz mentioned would make it much bigger. So we're likely going to need this query.